src/preprocessing/passes/ackermann.cpp

```diff
@@ -216,6 +324,23 @@ PreprocessingPassResult Ackermann::applyInternal(
         i, d_funcToSkolem.apply((*assertionsToPreprocess)[i]));
   }
 
+  /* the current version only support BV for removing uninterpreted sorts */
+  if (d_logic.isTheoryEnabled(theory::THEORY_BV))
```

Comment:
We should consider just always enabling `THEORY_BV` when Ackermannization is used. @yoni206 what do you think?

Reply:
Good question.
First, I think BV is only needed if Ackermannization is used *and* uninterpreted sorts are used.
Second, always translating uninterpreted sorts to BV sorts is a possibility. Another possibility, e.g., for QF_UFLIA, is to translate uninterpreted sorts to `Int` and enforce `distinct` constraints on all variables of that sort. The former has less constraints but involves theory combination, while the latter is the other way around. Do you think there is a clear winner? do you think it is interesting to try both?

Reply:
> Should we output an error if uninterpreted sorts are used but not eliminated by Ackermannization or should we just silently keep them?

Well... If the user explicitly turned on Ackermannization, then I would go for exception. But if Ackermannization was turned on by `bitblast=eager`, then I think we should silently keep them. Does this make sense?

> I think that it is sufficient to just turn the uninterpreted sorts into Ints, actually (since the cardinality of an uninterpreted sort is infinite).

In this example, there is a single uninterpreted sort `U`. But if you had `U1` and `U2`, shouldn't you assert that `u1 != u2` for every `u1 in U1` and `u2 in U2`?

Reply:
> But if Ackermannization was turned on by `bitblast=eager`, then I think we should silently keep them.

If that's the case, then we implicitly have BV because that's what the eager bit-blaster operates on :)

> In this example, there is a single uninterpreted sort U. But if you had U1 and U2, shouldn't you assert that u1 != u2 for every u1 in U1 and u2 in U2?

I guess that's the same argument that we have below about the different bit-widths. Assuming that you are right about the argument below, you'd have to divide `Int`s into ranges belonging to different sorts, I think.

Reply:
Doing `distinct` between integers that come from different sorts is a more general solution than dividing to ranges I think.

Reply:
Looking again at the PR, I think we should *not* throw an exception if BV is not enabled, but just leave the uninterpreted sorts.
This PP is called `ackermannization`, not `uninterpreted-sort-elimination`. If a user has a QF_UFLIA problem with uninterpreted sorts, she should be able to perform Ackermannization in order to eliminate function symbols, leaving uninterpreted sorts intact.
This makes me think that maybe uninterpreted-sort-elimination should be a separate preprocessing pass @4tXJ7f what do you think about that?

Either way, I think we should eliminate uninterpreted sorts only if BV is enabled. Otherwise, leave them in the formula.